accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -613,6 +606,198 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+a value of non-nullable type `T` immediately promotes the variable to the
+non-nullable type.*
+
+```dart
+void test() {
+  int? x = 3; // x is declared at `int?`
+  x.isEven; // Valid, x has been promoted to `int`
+  x = null; // Valid, demotes to the declared type.
+}
+```
+
+Local variables with no explicitly written type but with an initializer are
+given an inferred declared type equal to the type of their initializer.  In the
+case that the type of the initializer is a promoted type variable `X & T`, the
+inferred type of the variable shall be `X`.  However, such a variable shall be
+treated as immediately promoted to `X & T`.
+
+In the null safety release, local variables with no explicitly written type and
+no initializer are subject to further promotion based inference.  The variables
+under consideration here are those declared as `var x;`, `final x;`, `late var
+x;`, or `late final x;`.  We refer to such variables as **untyped variables**.
+
+An **initializing assignment** is an assignment to an **untyped variable** when
+that variable is in the **definitely unassigned** state.  Such an assignment on
+a path shall cause the variable in question to be treated by the flow analysis
+on that path as if it were declared with the assigned expression as its
+initializing expression, up until the next join point in the program.  We refer
+to this inferred type as the **initialization inferred type** of the variable.
+The **initialization inferred type** of the variable is always treated as a
+**type of interest** for the purposes of promotion (including its non-nullable
+version if applicable).
+
+```dart
+void test() {
+  var x; // x is an untyped variable
+  x = 3; // x is given the initialization inferred type of `int`
+}
+```
+
+At a join point in the program, it is an error if an **untyped variable** has
+been given an **initialization inferred type** on one of the paths to the join
+point and not another, unless the variable is declared as `late`.
+
+```dart
+void test(bool b) {
+  var x; // x is an untyped variable
+  if (b) {
+    x = 3;
+  } // Error, x has not been given an inferred type on all paths
+
+  late var y; // y is a late untyped variable
+  if (b) {
+    y = 3;
+  } // Ok, y has inferred type `int`
+}
+```
+
+At a join point in the program, if an **untyped variable** has been given an
+**initialization inferred type** on more than one of the paths to the join
+point, the variable is treated after the join point as having the
+**initialization inferred type** equal to the upper bound of the
+**initialization inferred types** from the paths into the join point.  The
+**initialization inferred type** so computed is treated as a **type of
+interest** for the variable in question (including its non-nullable version if
+applicable).  The promoted type chains for the variable on the respective paths
+are intersected as usual.
+
+```dart
+void test(bool b) {
+  var x; // x is an untyped variable
```

Comment:
So what is the "declared type" of [x] and [y] (below), i.e. the type we put on the variable declaration for backends to see? Is it `dynamic`?

Reply:
If tools expect all variables to have a declared type, then we might want to give them one, even though this variable actually has *no* declared type. (It might also make things easier for ourselves if we can consistently refer to the declared type of any variable).

It shouldn't matter what the type is, because it should not be exposed in any way. A RHS occurrence of `x` is a compile-time error by itself, so nobody needs to ask about its type.
A LHS occurrence of `x` doesn't need a type. Any type would be consistent.


I'd go for either `void` or `Never`, with `void` being the most consistent (because we can promote it to something else), but `Never` being totally true as well.

Reply:
@johnniwinther Good question.  I think there are two answers I can see using.  

I think the best answer is as follows:
  - If we insert an artificial exist node and join all variables on the exit node, then every variable which is given a initialization inferred type is eventually given a maximal initialization inferred type which is a super type of all of its initialization inferred types.  We treat that type as the declared type of the variable, and each initializing assignment is a promotion.
  - For variables which are never given an initialization inferred type (and hence must never be read), we give them the type `Object?` (or `Never`?).

The second best answer is that we just give all such variables the initial type `Object?`.  I think this is inferior for the purposes of error reporting and documentation in the analyzer, not sure how much difference it makes in CFE.

WDYT?

Reply:
Actually, thinking further about the above, it does raise a more general question.  Should this program be legal?

```dart
void test() {
  var x;
  if (something) {
    x = 3;
  } else if (somethingelse) {
    x = 3.0;
  } else {
     return;
  }
 print(x);
}
```

That is, should we allow variables to not be initialized on joins from unreachable paths?

Reply:
Yes.

The unreachable path should not have any effect on the join (its not joining into the flow at all). If it does, then we have an issue with our flow analysis.

Reply:
I have updated this to take reachability into account, and to specify a documentation type for untyped variables.  @johnniwinther @lrhn PTAL

Reply:
LGTM.